File: swift/Sources/DeviceAiLlm/Session/ChatConfig.swift
/// Configuration for a `ChatSession`.
///
/// Applied via a closure on session creation — same DSL pattern as Kotlin:
/// ```swift
/// // Swift
/// DeviceAI.llm.chat(modelPath: path) {
///     $0.systemPrompt = "You are helpful."
///     $0.temperature  = 0.8
/// }
///
/// // Kotlin equivalent
/// DeviceAI.llm.chat(path) {
///     systemPrompt = "You are helpful."
///     temperature  = 0.8f
/// }
/// ```
///
/// All fields have sensible defaults — zero-config session works out of the box.
/// Mirrors `ChatConfig` in kotlin/llm exactly.
public final class ChatConfig: @unchecked Sendable {

    // ── Conversation ──────────────────────────────────────────────────────────

    /// System prompt applied to every request. Shapes the model's persona.
    public var systemPrompt: String = "You are a helpful assistant."

    // ── Generation ────────────────────────────────────────────────────────────

    /// Maximum tokens to generate per response.
    public var maxTokens: Int    = 512

    /// Sampling temperature. 0 = deterministic, 1 = creative. Range: 0–2.
    public var temperature: Float = 0.7

    /// Nucleus sampling. Range: 0–1.
    public var topP: Float        = 0.9

    /// Top-k sampling. 0 = disabled.
    public var topK: Int          = 40

    /// Repetition penalty. Values > 1 discourage repeating phrases.
    public var repeatPenalty: Float = 1.1

    // ── Engine (init-time) ────────────────────────────────────────────────────

    /// KV-cache context window in tokens. Larger = more history, more RAM.
    public var contextSize: Int = 4096

    /// CPU threads for inference.
    public var threads: Int     = 4

    /// Use GPU (Metal). Default: true — Metal is 3–5× faster on Apple Silicon.
    public var useGpu: Bool     = true

    public init() {}
}

File: kotlin/llm/src/commonMain/kotlin/dev/deviceai/llm/ChatConfig.kt
package dev.deviceai.llm

/**
 * Configuration for a [ChatSession].
 *
 * Passed as a DSL block to [dev.deviceai.core.LlmModule.chat]:
 * ```kotlin
 * val session = DeviceAI.llm.chat(modelPath) {
 *     systemPrompt = "You are a helpful assistant."
 *     temperature  = 0.8f
 *     maxTokens    = 512
 * }
 * ```
 *
 * All fields have sensible defaults — a zero-config session works out of the box.
 */
class ChatConfig {

    // ── Conversation ─────────────────────────────────────────────────────────

    /**
     * System prompt applied to every request in this session.
     * Shapes the model's persona, tone, and behaviour.
     */
    var systemPrompt: String = "You are a helpful assistant."

    // ── Generation ───────────────────────────────────────────────────────────

    /** Maximum number of tokens to generate per response. */
    var maxTokens: Int = 512

    /**
     * Sampling temperature. Higher = more creative / random.
     * Lower = more deterministic / focused.
     * Range: 0.0–2.0. Default: 0.7.
     */
    var temperature: Float = 0.7f

    /**
     * Nucleus sampling threshold. Tokens whose cumulative probability
     * exceeds this value are discarded.
     * Range: 0.0–1.0. Default: 0.9.
     */
    var topP: Float = 0.9f

    /**
     * Top-k sampling. Considers only the k most probable tokens at each step.
     * Set to 0 to disable. Default: 40.
     */
    var topK: Int = 40

    /**
     * Repetition penalty. Values > 1.0 discourage repeating tokens.
     * Range: 1.0–2.0. Default: 1.1.
     */
    var repeatPenalty: Float = 1.1f

    // ── Engine (init-time) ────────────────────────────────────────────────────

    /**
     * Number of CPU threads for inference.
     * Default: 4.
     */
    var threads: Int = 4

    /**
     * Whether to use GPU acceleration — Metal on iOS, Vulkan on Android.
     * Default: true.
     */
    var useGpu: Boolean = true

    // ── Internal helpers ──────────────────────────────────────────────────────

    internal fun toInitConfig() = LlmInitConfig(
        maxThreads = threads,
        useGpu     = useGpu,
    )

    internal fun toGenConfig() = LlmGenConfig(
        maxTokens     = maxTokens,
        temperature   = temperature,
        topP          = topP,
        topK          = topK,
        repeatPenalty = repeatPenalty,
    )
}
